File: 0017-letter-combinations-of-a-phone-number.py
from typing import List

class Solution:
    convert_map = {
        "2": "abc",
        "3": "def",
        "4": "ghi",
        "5": "jkl",
        "6": "mno",
        "7": "pqrs",
        "8": "tuv",
        "9": "wxyz",
    }

    def letterCombinations(self, digits: str) -> List[str]:
        # 回溯
        result = []
        if not digits:
            return result

        def backtrack(conbination, nextdigit):
            if len(nextdigit) == 0:
                result.append(conbination)
            else:
                for letter in self.convert_map[nextdigit[0]]:
                    backtrack(conbination + letter,nextdigit[1:])

        backtrack('',digits)
        return result


    def __letterCombinations(self, digits: str) -> List[str]:
        # 队列
        if not digits:
            return []

        result = [""]

        for digit in digits:
            for _ in range(len(result)):
                tmp = result.pop(0)
                for letter in self.convert_map[digit]:
                    result.append(tmp + letter)

        return result

    def _letterCombinations(self, digits: str) -> List[str]:
        # 暴力四层循环，使用空字符做哨兵

        result = []
        if not digits:
            return result

        wait_to_use = [" "] * 4
        vertical_length = [0] * 4
        i = 0

        for d in digits:
            if d in "23456789":
                wait_to_use[i] = self.convert_map[d]
                vertical_length[i] = len(self.convert_map[d])
                i += 1

        for i in range(vertical_length[0]):
            for j in range(max(vertical_length[1], 1)):
                temp = wait_to_use[0][i]
                if wait_to_use[1][j] == " ":
                    result.append(temp)
                    break
                for k in range(max(vertical_length[2], 1)):
                    temp = wait_to_use[0][i] + wait_to_use[1][j]
                    if wait_to_use[2][k] == " ":
                        result.append(temp)
                        break
                    for l in range(max(vertical_length[3], 1)):
                        temp = wait_to_use[0][i] + wait_to_use[1][j] + wait_to_use[2][k]
                        if wait_to_use[3][l] == " ":
                            result.append(temp)
                            break
                        temp +=  wait_to_use[3][l]
                        result.append(temp)

        return result


if __name__ == "__main__":
    digits = "23"
    print(Solution().letterCombinations(digits))

File: 0017-letter-combinations-of-a-phone-number.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

void back_track(char *converter[], char **result, char * combination, char * digits, int* index, int com_length, int length){
    if (*digits == '\0') {
        // result[*index] = (char*)malloc(sizeof(char) * com_length);
        result[*index] = combination;
        (*index)++;
    }else {
        // int c_i = 0;
        // while (*(conbination + c_i) != '\0') {
        //     c_i += 1;
        // }
        char head_ch = digits[0];
        char * letters = converter[head_ch - '2'];
        while(*letters != '\0') {
            char * next_combination = (char *)malloc(sizeof(char) * length);
            strcpy(next_combination, combination);
            next_combination[com_length] = *letters;
            next_combination[com_length + 1] = '\0';
            char * next_digits = (char *)malloc(sizeof(char) * length);
            strcpy(next_digits, digits + 1);
            back_track(converter, result, next_combination, next_digits, index, com_length + 1, length);
            letters += 1;
        }
    }
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
char ** letterCombinations(char * digits, int* returnSize){
    char *converter[] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    int letter_length[] = {3, 3, 3, 3, 3, 4, 3, 4};
    *returnSize = 1;
    int length = 0;
    while (*(digits + length) != '\0') {
        *returnSize *= letter_length[*(digits + length) - '2'];
        length++;
    }
    char ** result = (char **)malloc(sizeof(char *)*(*returnSize));
    char * conbination = (char*)malloc(length*sizeof(char));
    int * initial_index = (int *)malloc(sizeof(int));
    *initial_index = 0;
    back_track( converter, result, conbination, digits, initial_index, 0, length);
    return result;
}


int main(int argc, char *argv[]) {
    int *returnSize = (int *)malloc(sizeof(int));
    char ** result = letterCombinations("23", returnSize);
    for (int i = 0; i < *returnSize ; i++) {
        printf("%s\n", result[i]);
    }
    return 0;
}
